

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Create datasets &mdash; VoSeq 2.0.0-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../static/favicon.ico"/>
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="VoSeq 2.0.0-beta documentation" href="../index.html"/>
        <link rel="next" title="Search" href="search.html"/>
        <link rel="prev" title="Create taxonset" href="create-taxonset.html"/> 

  
  <script src="../static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> VoSeq
          

          
          </a>

          
            
            
              <div class="version">
                2.0.0-beta
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../news.html">News</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../intro/overview.html">VoSeq overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro/install.html">Download to install VoSeq</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro/install.html#installation-guide">Installation guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro/administrate-server.html">Administrate the server with your installation of VoSeq</a></li>
<li class="toctree-l1"><a class="reference internal" href="adding-vouchers.html">Adding vouchers</a></li>
<li class="toctree-l1"><a class="reference internal" href="adding-genes.html">Adding genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="adding-sequences.html">Adding sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="create-taxonset.html">Create taxonset</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Create datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="search.html">Search</a></li>
<li class="toctree-l1"><a class="reference internal" href="upload-voucher-photos.html">Upload voucher photos</a></li>
<li class="toctree-l1"><a class="reference internal" href="create-voucher-table.html">Create Voucher table</a></li>
<li class="toctree-l1"><a class="reference internal" href="update-voucher.html">Update voucher</a></li>
<li class="toctree-l1"><a class="reference internal" href="../update-voseq.html">How to update Voseq</a></li>
<li class="toctree-l1"><a class="reference internal" href="../migrate-mysql.html">Migrate VoSeq&#8217;s MySQL database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../plugins.html">Plugins</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../thanks.html">Thanks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dev/development.html">Development of VoSeq</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">VoSeq</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Create datasets</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../sources/usage/create-datasets.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="create-datasets">
<span id="id1"></span><h1>Create datasets<a class="headerlink" href="#create-datasets" title="Permalink to this headline">Â¶</a></h1>
<p>We believe that one of VoSeq&#8217;s important features is the <strong>capability to create
dataset files of molecular sequences that are ready-to-run in phylogenetic
software</strong> such as <strong>MrBayes, TNT, PAUP, NEXUS</strong>, etc.</p>
<p>Now that you have voucher and sequence data in your installation of VoSeq,
you might want to create datasets for analysis of sequences in phylogenetic software.</p>
<p>In the <strong>user interface</strong>, you will find under the <strong>Tools</strong> section the link
<strong>Create new dataset</strong>. You will be shown a page to select the sequences you
want by entering the <strong>voucher codes</strong> and <strong>gene codes</strong>. You can select your
data to be in several formats (FASTA, NEXUS, etc), choose between codon positions,
as well as choosing what information your taxon names should include.</p>
<p>This will create a <strong>ready-to-use</strong> data set for analyses!</p>
<a class="reference internal image-reference" href="../images/create_dataset.png"><img alt="../images/create_dataset.png" class="align-center" src="../images/create_dataset.png" style="width: 600px;" /></a>
<p>The <strong>Outgroup</strong> field, if needed, should include the voucher code for the chosen outgroup taxa.</p>
<dl class="docutils">
<dt>Codon positions</dt>
<dd>Marking <strong>1st</strong>, <strong>2nd</strong> or <strong>3rd</strong> and unmarking <strong>all</strong> positions will
create a dataset with only the chosen position(s) for all genes.
<strong>Special</strong> will take you to a new page where you will be able to choose
which codon positions to include for each gene.
Marking <strong>amino acids</strong> will tranlate <strong>protein-coding</strong> genes with a set
<strong>genetic code</strong> , the others will be treated as normal dna, i.e. making
&#8220;mixed&#8221; datatype in Nexus for MrBayes, and setting partitions correctly in Nexus and PHYLIP format.
Note that codon position choices as well as translation to aminoacids are
only able to function if the user have specified a <strong>reading frame</strong> for
the chosen gene(s) (see <a class="reference internal" href="adding-genes.html#adding-genes"><span class="std std-ref">Adding genes</span></a>).</dd>
<dt>Partition by (position)</dt>
<dd>Here you can choose how to do partitioning for each gene.
<strong>as one</strong> will create one partition per gene, regardless of which codon
positions you include.
<strong>each</strong> will create a partition per codon position, whereas <strong>1st-2nd,
3rd</strong> will create one combined partition for the 1st and 2nd positions and
one separate for the 3rd codon positions.
Note that <strong>each</strong> and <strong>all</strong> are only possible to process with a per gene
specified <strong>reading frame</strong> (see <a class="reference internal" href="adding-genes.html#adding-genes"><span class="std std-ref">Adding genes</span></a>).</dd>
</dl>
<p>You can also chose to <strong>omit taxa from a taxonset that contains less than a
specific number sequences</strong>. Say you have a 10 gene data set and want to remove
all taxa with 5 or less! Easy! Just eneter a minimum number of genes!</p>
<p>If you have introns in your alignment you can choose to include or remove them
from the output data set. If included they will be treated as separate data
blocks and partitions for the Nexus and PHYLIP outputs!</p>
<p>The voucher codes can be entered one by one (separated by return) in the text
area or you could create a <a class="reference internal" href="create-taxonset.html#create-taxonset"><span class="std std-ref">Create taxonset</span></a> (a list of voucher codes for
a specific project).</p>
<a class="reference internal image-reference" href="../images/create_dataset2.png"><img alt="../images/create_dataset2.png" class="align-center" src="../images/create_dataset2.png" style="width: 600px;" /></a>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">As of version 1.5.0, protein-coding ability, aligned or not, introns
and genetic code will be set for each gene/alignment in the admin gene section!</p>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="search.html" class="btn btn-neutral float-right" title="Search" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="create-taxonset.html" class="btn btn-neutral" title="Create taxonset" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2013-2015, Carlos PeÃ±a, Tobias Malm.
      Last updated on Apr 09, 2017.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'2.0.0-beta',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: ''
        };
    </script>
      <script type="text/javascript" src="../static/jquery.js"></script>
      <script type="text/javascript" src="../static/underscore.js"></script>
      <script type="text/javascript" src="../static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>